src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs

```diff
@@ -41,69 +46,108 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
             var position = context.Span.Start;
             var document = context.Document;
             var cancellationToken = context.CancellationToken;
-
             var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
             var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-
             var token = root.FindToken(position);
-            var parameterNode = GetParameterNode(token, position);
-            if (parameterNode == null)
-            {
-                return;
-            }
+            var firstparameterNode = GetParameterNode(token, position);
+
+            // Creating a list and iterating it from the allows the right order or ifs
+            var listOfParameterNodes = GetParameterList(firstparameterNode);
+            var counter = 0;
 
-            // Only offered when there isn't a selection, or the selection exactly selects
-            // a parameter name.
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
-            if (!context.Span.IsEmpty)
+            // List with parameterNodes that pass all checks
+            var listOfValidParametersNodes = new List<SyntaxNode>();
+            var checkIfToReturn = false;
+
+            if (listOfParameterNodes != null)
             {
-                var parameterName = syntaxFacts.GetNameOfParameter(parameterNode);
-                if (parameterName == null || parameterName.Value.Span != context.Span)
+                foreach (var parameterNode in listOfParameterNodes)
+                {
+                    counter++;
+                    if (parameterNode == null)
+                    {
+                        continue;
+                    }
+
+                    // Only offered when there isn't a selection, or the selection exactly selects a parameter name.
```

Comment:
Some of the checks on selection to decide weather to offer refactoring do not apply to "add for all parameter" option. For example, do we want to offer "add for all" when selection exactly selects a parameter name? We need to figure out exactly when to offer this new refactoring.

Reply:
Also, these available checks don't determine whether a parameter is "valid" for refactoring, I believe we have logic in `ParameterValidForNullCheck` that does that, right? Seems to me that this loop is not needed, we can just pass entire parameter list to AddParameterCheck provider instead.

---
In reply to: [300806544](https://github.com/dotnet/roslyn/pull/36931#discussion_r300806544) [](ancestors = 300806544)

Reply:
True. but some checks are also made here. For example It checks the method symbol of each parameter, and whether it or its name are null. They don't determine whether a parameter is "valid" for refactoring - but that help determine which parameters are NOT valid. Let me know what you think - I could be wrong.

Reply:
I'm currently working on making refactorings offerings more consistent #35525 and:  

1) We have a set of helpers living in `IRefactoringHelpersService` that you can ask "here's my selection/location and I'm interested in this type (ParameterSyntax / ParameterListSyntax in your case), give me the SyntaxNode if refactoring for it should be offered / null if not".  Example usage can be seen here: https://github.com/dotnet/roslyn/blob/6dda10231bbfd873877f2e9775ece2bdf53d2889/src/Features/Core/Portable/InitializeParameter/AbstractInitializeParameterCodeRefactoringProvider.cs#L50 The `AbstractInitializeParameterCodeRefactoring` has actually been moved to that helper (not sure why it doesn't show in diffs, it's already merged in master). It would be great if we could work together so that it keeps using the helper (if there're any unsatisfied requirements we can extend it) and doesn't regress to custom handling of `context.Span`.

2) Generally our guidelines that would be applicable we arrived to (happy to revise them) were:
   - If user hasn't selected anything and is anywhere inside the header (in any argument) -> show Add check for all parameters (we consider the caret being in a header of ArgumentList).
   - If user selected less then one/one parameter -> show Add check only for that parameter (they selected -> they want that one specifically)
   - If user selected the whole parameter list -> show Add check to all (obvious)
    - Also, since argument initializers can be only primitive values, we came to a conclusion that the initializer should be part of the header (and thus applicable caret location to invoke the refactoring).

Reply:
3) As for concrete implementation, I'm slightly confused (because the diff here doesn't seem to be up to date) but I think it could look something like this: 
```csharp
var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();
var parameterNode = await refactoringHelperService.TryGetSelectedNodeAsync<TParameterSyntax>(document, context.Span, cancellationToken).ConfigureAwait(false);

// A concrete parameter node can be identified & it was a selection of one parameter (-> intentional)
if (parameterNode != null && !context.Span.IsEmpty)
{
   // Initialize just for that one parameterNode
}

var parameterListNode = parameterNode?.Parent ?? await refactoringHelperService.TryGetSelectedNodeAsync<TParameterListSyntax>(document, context.Span, cancellationToken).ConfigureAwait(false);

if (parameterListNode == null)
{
    return;
}

// Offer initialize (add null checks) for all parameters in parameterListNode 
```